Make StridedArray display as StridedArray

mbauman opened the pull request:
Instead of a huge union.  This is just a special-casing for this particular union instead of pursuing the general (and hard\!) typealias search.  But given the ubiquity of StridedArray in method lists, I find this special case to be warranted and hugely beneficial. Here is a sampling from `methods(*)`:

```julia
[242] *(A::LinearAlgebra.AbstractQ, b::StridedArray{T, 1} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:556
[243] *(A::LinearAlgebra.AbstractQ, B::StridedArray{T, 2} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:568
[244] *(Q::LinearAlgebra.AbstractQ, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:(StridedVecOrMat{T} where T) where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:623
[245] *(A::StridedArray{T, 2} where T, Q::LinearAlgebra.AbstractQ) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:668
[246] *(A::StridedArray{T, 2} where T, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:LinearAlgebra.AbstractQ where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:708
[247] *(A::LinearAlgebra.LQPackedQ, B::StridedVecOrMat{T} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:153
[248] *(A::LinearAlgebra.LQPackedQ, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:(StridedVecOrMat{T} where T) where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:177
[249] *(A::StridedVecOrMat{T} where T, adjQ::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:LinearAlgebra.LQPackedQ where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:216
[250] *(A::StridedVecOrMat{T} where T, Q::LinearAlgebra.LQPackedQ) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:240
```

mbauman committed:
Make StridedArray display as StridedArray

Instead of a huge union.  This is just a special-casing for this particular union instead of pursuing the general (and hard\!) typealias search.  But given the ubiquity of StridedArray in method lists, I find this special case to be warranted and hugely beneficial. Here is a sampling from `methods(*)`:

```julia
[242] *(A::LinearAlgebra.AbstractQ, b::StridedArray{T, 1} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:556
[243] *(A::LinearAlgebra.AbstractQ, B::StridedArray{T, 2} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:568
[244] *(Q::LinearAlgebra.AbstractQ, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:(StridedVecOrMat{T} where T) where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:623
[245] *(A::StridedArray{T, 2} where T, Q::LinearAlgebra.AbstractQ) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:668
[246] *(A::StridedArray{T, 2} where T, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:LinearAlgebra.AbstractQ where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/qr.jl:708
[247] *(A::LinearAlgebra.LQPackedQ, B::StridedVecOrMat{T} where T) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:153
[248] *(A::LinearAlgebra.LQPackedQ, adjB::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:(StridedVecOrMat{T} where T) where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:177
[249] *(A::StridedVecOrMat{T} where T, adjQ::LinearAlgebra.Adjoint{#s614,#s613} where #s613<:LinearAlgebra.LQPackedQ where #s614) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:216
[250] *(A::StridedVecOrMat{T} where T, Q::LinearAlgebra.LQPackedQ) in LinearAlgebra at /home/mbauman/julia/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/lq.jl:240
```

mbauman commented:
Interesting — the AppVeyor failures are actually real here.  I'll have to scrounge up a windows VM to figure out what's going on there.  The failure is `startswith(sprint(show, StridedVecOrMat), "StridedVecOrMat")`.

Travis Mac is a network connectivity timeout.

mbauman commented:
Alright, now the AppVeyor failures are unrelated.

StefanKarpinski commented:
Let's do this! I propose merging if there are no objections by triage.

Keno commented on base/show.jl:
```diff
@@ -425,6 +425,18 @@ function show(io::IO, @nospecialize(x::Type))
         return
     elseif x isa Union
+        if x <: StridedArray && x.a <: DenseArray
+            T, N = x.a.parameters
```
Note: Not guaranteed to be available from the subtype check alone

JeffBezanson commented on base/show.jl:
```diff
@@ -425,6 +425,18 @@ function show(io::IO, @nospecialize(x::Type))
         return
     elseif x isa Union
+        if x <: StridedArray && x.a <: DenseArray
+            T, N = eltype(x.a), ndims(x.a)
```
I think this `ndims` call will fail if `x.a` is an abstract type.

mbauman replied:
Ah, good point. What if I just add an `applicable(ndims, x.a)` to the condition above?

mbauman replied:
Fixed by explicitly ensuring that `x.a` is explicitly a `DenseArray` and not just a subtype on the previous line with:

```Julia
if x.a isa DataType && Core.Compiler.typename(x.a) === Core.Compiler.typename(DenseArray)
    T, N = x.a.parameters
```

Then I can dig into `.parameters` and know they mean `T, N`.

StefanKarpinski commented:
Triage approves this as a bandaid until we have a more general solution to the excessively verbose printing of complex types.

StefanKarpinski commented:
NEWSworthy?

vtjnash commented:
Can/should this use `showarg` instead?

mbauman commented:
Oh cool. I had forgotten about `showarg`. Yes, that looks to accomplish what I want without making this totally obscure what a `StridedArray` is.

mbauman commented:
Oh, nope, `showarg` isn't going to do the trick. It works in the value domain, not in the type domain.

mbauman commented:
Rebased, fixed up, and I further added support for `Union{StridedArray, SomethingElse}` because it was easy to do and I noticed the crazy `StridedOrTriangularMatrix` Union in `methods(*)`:

https://github.com/JuliaLang/julia/blob/7d27fa883ca06e34863e90cae72d8e7438c108ae/stdlib/SparseArrays/src/linalg.jl#L7
